modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/SearchTemplateRequest.java

```diff
@@ -134,6 +161,62 @@ public ActionRequestValidationException validate() {
         return validationException;
     }
 
+    private static ParseField ID_FIELD = new ParseField("id");
+    private static ParseField SOURCE_FIELD = new ParseField("source", "inline", "template");
+
+    private static ParseField PARAMS_FIELD = new ParseField("params");
+    private static ParseField EXPLAIN_FIELD = new ParseField("explain");
+    private static ParseField PROFILE_FIELD = new ParseField("profile");
+
+    private static final ObjectParser<SearchTemplateRequest, Void> PARSER;
+    static {
+        PARSER = new ObjectParser<>("search_template");
+        PARSER.declareField((parser, request, s) ->
+                request.setScriptParams(parser.map())
+            , PARAMS_FIELD, ObjectParser.ValueType.OBJECT);
+        PARSER.declareString((request, s) -> {
+            request.setScriptType(ScriptType.STORED);
+            request.setScript(s);
+        }, ID_FIELD);
+        PARSER.declareBoolean(SearchTemplateRequest::setExplain, EXPLAIN_FIELD);
+        PARSER.declareBoolean(SearchTemplateRequest::setProfile, PROFILE_FIELD);
+        PARSER.declareField((parser, request, value) -> {
+            request.setScriptType(ScriptType.INLINE);
+            if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+                //convert the template to json which is the only supported XContentType (see CustomMustacheFactory#createEncoder)
+                try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+                    request.setScript(Strings.toString(builder.copyCurrentStructure(parser)));
+                } catch (IOException e) {
+                    throw new ParsingException(parser.getTokenLocation(), "Could not parse inline template", e);
+                }
+            } else {
+                request.setScript(parser.text());
+            }
+        }, SOURCE_FIELD, ObjectParser.ValueType.OBJECT_OR_STRING);
+    }
+
+    public static SearchTemplateRequest fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, new SearchTemplateRequest(), null);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+
+        if (scriptType == ScriptType.STORED) {
```

Comment:
super minor nit: maybe this could be a switch ?

Reply:
I would usually do use a switch, but here I found it harder to read because the enum values must be unqualified (`case STORED` instead of `case ScriptType.STORED`).